docs/deployment.md

```diff
@@ -27,225 +28,308 @@ Configuration of the PDP is accomplished through two mechanisms:
+    another one.
+    
+    1. Set up user mapping in your local Docker daemon. 
+        Follow the instructions in the `pdp-docker` documentation:
+         [Running tests and PDP application on workstation - Procedure](https://github.com/pacificclimate/pdp-docker#procedure). You will, 
+         unfortunately, need to clone 
+         [`pdp-docker`](https://github.com/pacificclimate/pdp-docker) 
+         in order to obtain the shell script you will run.
+           
+         This procedure
+         - establishes a user on your workstation named `dockeragent`,
+         - makes it a member of your own user group so that it has access to
+           your filesystem, and
+         - sets up your Docker daemon to map user `dockeragent` inside containers
+         onto `dockeragent` in your workstation. 
+    
+    1. Create `docker/local-run/common-with-passwords.env` from 
+    `docker/local-run/common.env` by adding passwords for the `pcic_meta` and 
+    `crmp` databases.
+    
+    1. Create `docker/local-run/pgbounce_users-with-passwords.txt` from  
+    `docker/dev-local/pgbounce_users.txt` by inserting correct md5 sums.
+    
+    1. Edit your `/etc/hosts`: Add `pdp.localhost` to the line starting with 
+    `127.0.0.1`. The result will look like 
+    
+        ```
+        127.0.0.1       localhost pdp.localhost
+        ```
+        This allows the Nginx reverse proxy set up by the docker-compose 
+        to refer to the domain `pdp.localhost`. Note that the frontend container is 
+        configured with `APP_ROOT` and `DATA_ROOT` using this domain.
+
+1. **Build the image**
+
+    The image need only be (re)built when:
+    
+    1. the project is first cloned, or
+    2. any of the `*requirements.txt` files change, or
+    3. the local-run Dockerfile changes.
+    
+    The built image contains all dependencies specified in those files (but not 
+    the PDP codebase). It forms the basis for installing and running your local 
+    codebase.
+    
+    To build the image:
+    
+    ```
+    docker-compose -f docker/local-run/docker-compose.yaml build
+    ```
+    
+    Image build can take several minutes.
+
+1. **Mount the gluster `/storage` volume**
+   
+    Mount locally to `/storage` so that those data files are accessible on 
+    your workstation.
+    
+    ```
+    sudo mount -t cifs -o username=XXXX@uvic.ca //pcic-storage.pcic.uvic.ca/storage/ /storage
+    ```
+
+1. **Start the containers**
+
+    ```
+    docker-compose -f docker/dev-local/docker-compose.yaml up -d
+    ```
+    
+    This starts containers for the backend, frontend, pgbouncer, and a local 
+    reverse proxy that maps the HTTP addresses `pdp.localhost:5000` onto the 
+    appropriate containers (mainly to avoid CORS problems).
+    
+    The frontend and backend containers automatically start up the application.
+
+1. **Point your browser at `pdp.localhost:5000/portal/<name>/map/`**
+
+    This should load the named PDP portal.
```

Comment:
The frontend and backend containers failed to start for me, with the following lightly edited log messages:

```
[python 2 deprecation warning]
[requirements for many packages are already satisfied]
Collecting urllib3<1.25,>=1.21.1
  Downloading urllib3-1.24.3-py2.py3-none-any.whl (118 kB)
     |################################| 118 kB 3.4 MB/s 
[some other package requirements are already satisfied]
Building wheels for collected packages: pydap.responses.xls
  Building wheel for pydap.responses.xls (setup.py) ... done
  Created wheel for pydap.responses.xls: filename=pydap.responses.xls-0.1-py2-none-any.whl size=4222 sha256=0267fba60a25bb0178b8f8a93fef20bc27b1bc0afcd45d46145670f1ede9c2b3
  Stored in directory: /opt/dockeragent/.cache/pip/wheels/b3/28/7c/55b72cd9105cb4f9af5f4f5487c2c7381b747637f5e04b14fb
Successfully built pydap.responses.xls
Installing collected packages: pydap.responses.xls, pdp, urllib3
  Running setup.py develop for pdp
    ERROR: Command errored out with exit status 1:
     command: /usr/bin/python -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/codebase/setup.py'"'"'; __file__='"'"'/codebase/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' develop --no-deps --user --prefix=
         cwd: /codebase/
    Complete output (4 lines):
    running develop
    running egg_info
    creating pdp.egg-info
    error: could not create 'pdp.egg-info': Permission denied
    ----------------------------------------
ERROR: Command errored out with exit status 1: /usr/bin/python -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/codebase/setup.py'"'"'; __file__='"'"'/codebase/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' develop --no-deps --user --prefix= Check the logs for full command output.
[some successful gunicorn startup messages]
2021-07-19 02:49:47 [39] [ERROR] Exception in worker process
Traceback (most recent call last):
  File "/opt/dockeragent/.local/lib/python2.7/site-packages/gunicorn/arbiter.py", line 583, in spawn_worker
    worker.init_process()
  File "/opt/dockeragent/.local/lib/python2.7/site-packages/gunicorn/workers/ggevent.py", line 203, in init_process
    super(GeventWorker, self).init_process()
  File "/opt/dockeragent/.local/lib/python2.7/site-packages/gunicorn/workers/base.py", line 129, in init_process
    self.load_wsgi()
  File "/opt/dockeragent/.local/lib/python2.7/site-packages/gunicorn/workers/base.py", line 138, in load_wsgi
    self.wsgi = self.app.wsgi()
  File "/opt/dockeragent/.local/lib/python2.7/site-packages/gunicorn/app/base.py", line 67, in wsgi
    self.callable = self.load()
  File "/opt/dockeragent/.local/lib/python2.7/site-packages/gunicorn/app/wsgiapp.py", line 52, in load
    return self.load_wsgiapp()
  File "/opt/dockeragent/.local/lib/python2.7/site-packages/gunicorn/app/wsgiapp.py", line 41, in load_wsgiapp
    return util.import_app(self.app_uri)
  File "/opt/dockeragent/.local/lib/python2.7/site-packages/gunicorn/util.py", line 350, in import_app
    __import__(module)
  File "/opt/dockeragent/.local/lib/python2.7/site-packages/gevent/builtins.py", line 96, in __import__
    result = _import(*args, **kwargs)
  File "/codebase/pdp/__init__.py", line 16, in <module>
    from pdp.portals import updateConfig
  File "/opt/dockeragent/.local/lib/python2.7/site-packages/gevent/builtins.py", line 96, in __import__
    result = _import(*args, **kwargs)
  File "/codebase/pdp/portals/__init__.py", line 4, in <module>
    from pdp_util.raster import db_raster_configurator, RasterServer, RasterCatalog
  File "/opt/dockeragent/.local/lib/python2.7/site-packages/gevent/builtins.py", line 96, in __import__
    result = _import(*args, **kwargs)
  File "/opt/dockeragent/.local/lib/python2.7/site-packages/pdp_util/raster.py", line 5, in <module>
    from pydap.wsgi.app import DapServer
  File "/opt/dockeragent/.local/lib/python2.7/site-packages/gevent/builtins.py", line 96, in __import__
    result = _import(*args, **kwargs)
  File "/opt/dockeragent/.local/lib/python2.7/site-packages/pydap/wsgi/app.py", line 40, in <module>
    from pydap.handlers.lib import get_handler
  File "/opt/dockeragent/.local/lib/python2.7/site-packages/gevent/builtins.py", line 96, in __import__
    result = _import(*args, **kwargs)
  File "/opt/dockeragent/.local/lib/python2.7/site-packages/pydap/handlers/lib.py", line 44, in <module>
    class BaseHandler(object):
  File "/opt/dockeragent/.local/lib/python2.7/site-packages/pydap/handlers/lib.py", line 55, in BaseHandler
    responses = load_responses()
  File "/opt/dockeragent/.local/lib/python2.7/site-packages/pydap/responses/lib.py", line 8, in load_responses
    return dict((r.name, r.load()) for r in iter_entry_points('pydap.response'))
  File "/opt/dockeragent/.local/lib/python2.7/site-packages/pydap/responses/lib.py", line 8, in <genexpr>
    return dict((r.name, r.load()) for r in iter_entry_points('pydap.response'))
  File "/opt/dockeragent/.local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2442, in load
    self.require(*args, **kwargs)
  File "/opt/dockeragent/.local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2465, in require
    items = working_set.resolve(reqs, env, installer, extras=self.extras)
  File "/opt/dockeragent/.local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 791, in resolve
    raise VersionConflict(dist, req).with_context(dependent_req)
ContextualVersionConflict: (urllib3 1.26.5 (/opt/dockeragent/.local/lib/python2.7/site-packages), Requirement.parse('urllib3<1.25,>=1.21.1'), set(['requests']))
2021-07-19 02:49:47 [39] [ERROR] Exception in worker process
Traceback (most recent call last):
  File "/opt/dockeragent/.local/lib/python2.7/site-packages/gunicorn/arbiter.py", line 583, in spawn_worker
    worker.init_process()
  File "/opt/dockeragent/.local/lib/python2.7/site-packages/gunicorn/workers/ggevent.py", line 203, in init_process
    super(GeventWorker, self).init_process()
  File "/opt/dockeragent/.local/lib/python2.7/site-packages/gunicorn/workers/base.py", line 129, in init_process
    self.load_wsgi()
  File "/opt/dockeragent/.local/lib/python2.7/site-packages/gunicorn/workers/base.py", line 138, in load_wsgi
    self.wsgi = self.app.wsgi()
  File "/opt/dockeragent/.local/lib/python2.7/site-packages/gunicorn/app/base.py", line 67, in wsgi
    self.callable = self.load()
  File "/opt/dockeragent/.local/lib/python2.7/site-packages/gunicorn/app/wsgiapp.py", line 52, in load
    return self.load_wsgiapp()
  File "/opt/dockeragent/.local/lib/python2.7/site-packages/gunicorn/app/wsgiapp.py", line 41, in load_wsgiapp
    return util.import_app(self.app_uri)
  File "/opt/dockeragent/.local/lib/python2.7/site-packages/gunicorn/util.py", line 350, in import_app
    __import__(module)
  File "/opt/dockeragent/.local/lib/python2.7/site-packages/gevent/builtins.py", line 96, in __import__
    result = _import(*args, **kwargs)
  File "/codebase/pdp/__init__.py", line 16, in <module>
    from pdp.portals import updateConfig
  File "/opt/dockeragent/.local/lib/python2.7/site-packages/gevent/builtins.py", line 96, in __import__
    result = _import(*args, **kwargs)
  File "/codebase/pdp/portals/__init__.py", line 4, in <module>
    from pdp_util.raster import db_raster_configurator, RasterServer, RasterCatalog
  File "/opt/dockeragent/.local/lib/python2.7/site-packages/gevent/builtins.py", line 96, in __import__
    result = _import(*args, **kwargs)
  File "/opt/dockeragent/.local/lib/python2.7/site-packages/pdp_util/raster.py", line 5, in <module>
    from pydap.wsgi.app import DapServer
  File "/opt/dockeragent/.local/lib/python2.7/site-packages/gevent/builtins.py", line 96, in __import__
    result = _import(*args, **kwargs)
  File "/opt/dockeragent/.local/lib/python2.7/site-packages/pydap/wsgi/app.py", line 40, in <module>
    from pydap.handlers.lib import get_handler
  File "/opt/dockeragent/.local/lib/python2.7/site-packages/gevent/builtins.py", line 96, in __import__
    result = _import(*args, **kwargs)
  File "/opt/dockeragent/.local/lib/python2.7/site-packages/pydap/handlers/lib.py", line 44, in <module>
    class BaseHandler(object):
  File "/opt/dockeragent/.local/lib/python2.7/site-packages/pydap/handlers/lib.py", line 55, in BaseHandler
    responses = load_responses()
  File "/opt/dockeragent/.local/lib/python2.7/site-packages/pydap/responses/lib.py", line 8, in load_responses
    return dict((r.name, r.load()) for r in iter_entry_points('pydap.response'))
  File "/opt/dockeragent/.local/lib/python2.7/site-packages/pydap/responses/lib.py", line 8, in <genexpr>
    return dict((r.name, r.load()) for r in iter_entry_points('pydap.response'))
  File "/opt/dockeragent/.local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2442, in load
    self.require(*args, **kwargs)
  File "/opt/dockeragent/.local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2465, in require
    items = working_set.resolve(reqs, env, installer, extras=self.extras)
  File "/opt/dockeragent/.local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 791, in resolve
    raise VersionConflict(dist, req).with_context(dependent_req)
ContextualVersionConflict: (urllib3 1.26.5 (/opt/dockeragent/.local/lib/python2.7/site-packages), Requirement.parse('urllib3<1.25,>=1.21.1'), set(['requests']))
2021-07-19 02:49:47 [39] [INFO] Worker exiting (pid: 39)
2021-07-19 02:49:47 [39] [INFO] Worker exiting (pid: 39)
2021-07-19 02:49:47 [35] [INFO] Shutting down: Master
2021-07-19 02:49:47 [35] [INFO] Shutting down: Master
2021-07-19 02:49:47 [35] [INFO] Reason: Worker failed to boot.
2021-07-19 02:49:47 [35] [INFO] Reason: Worker failed to boot.
```

Reply:
This seems likely to be a failing in the container to host user mapping I set up. Let's confer on this later. Meantime I will proceed.